### DISCUSSION TITLE
Leviton Z-Wave devices firmware upgade.

### DISCUSSION THREAD

**Discussion by ChrisLizon** [zwave-js/zwave-js#4087] [Q&A] [answered]
Has anyone had a chance to try the firmware update tool with Leviton devices? I have 6 DZ1KD at my parent's house that are on firmware 1.13. and two DZ15S on 1.18. They tend to show the wrong status (on vs off, or always off) and apparently they fixed this in firmware 1.20. I always attributed it to their Wink hub being... well.. a Wink hub, but now that I have them on ZwaveJs2Mqtt/Home Assistant I would like to either fix the devices or replace them.

The updates are available from [Leviton here](https://hubconnecteddecorasmart.zendesk.com/hc/en-us/articles/4407521680155-Can-Leviton-Z-Wave-Products-Firmware-Be-Updated-) 

I was able to extract a 2kb binary file from the installer using NirSoft's ResourcesExtract. ZwaveJS2MQTT showed that 30 segments were uploaded, but even after air gapping the switch it still seems that ZwaveJS is seeing firmware 1.13.

I know @djdizzyd was working on a firmware updater for Hubitat and did some experimenting with trying to pull the firmware out of the update tool in Intel Hex format, but there was something wrong with the files and I think that effort was abandoned. When I run the ZwaveJS update command its spits out

```
2022-01-14 14:35:42.372 INFO ZWAVE: Calling api beginFirmwareUpdate with args: [
7,
'Leviton-DZ1KD_01_20.hex',
<Buffer 3a 30 32 30 30 30 30 30 34 30 30 30 30 46 41 0d 0a 3a 32 30 30 30 30 30 30 30 30 44 37 37 46 39 37 36 36 41 41 30 30 45 34 45 46 37 41 33 36 46 35 43 ... 315372 more bytes>,
0,
[length]: 4
]
2022-01-14 14:35:42.407 INFO ZWAVE: Unable to extract firmware from file: Duplicated data at record 2050 (:200000000DE45302DF938868A94B2A718BE5AA07A03C19ADD7BBB8267249CEE3C21987F2B6) beginFirmwareUpdate undefined
```
checking in a text editor it looks like the hex addresses start over at line 2050, so I'm not sure if that's a issue with the file or if it's not being understood by the tool. 
![image](https://user-images.githubusercontent.com/1197737/149574886-92f2169a-f9b5-4413-bafc-519f747f2163.png)

They are using the Zooz ZST10 which isn't compatible with the Leviton updater files, but I may end up just taking my Aeotec stick to my parent's house, do the exclusion/inclusion/update/exclusion/inclusion song and dance. But I'd be willing to do some experimenting if anyone sees value in making this functionality available for others.

**Comment by ChrisLizon** [comment, marked as answer]
Alright just in case anyone is following along at home I'll give you my update. 

I spent the afternoon at my Parent's house and brought my Aeotec z-stick gen 5 because it was compatible with Leviton's updater software. Turns out that software is a steaming pile of garbage. I ended up pairing the device using the Silcon Labs Z-wave PC Controller, so I know the stick was working, but the Leviton software just never worked and showed no appreciable status indicator.

In frustration I tossed down the $20 for the HomeSeer Z-Flash package and that one picked up my test device no problem. The software is actually very nice. While it was flashing I poked around and realized that the installer package extracts all of the firmware files to the install directory in c:\Program Files. I grabbed the appropriate .hec files and upgraded the other 7 devices with ZwaveJStoMQTT web interface. While a few of them were slower than others they all updated and the ZwaveJS2MQTT now shows the correct firmware version for all the Leviton devices in the house. They are reporting their status properly and Alexa no longer complains about the devices being unreachable so everyone is happy. I did have a weird experience where a bunch of devices stopped responding after the longest running device update, but restarting the ZwaveJS2MQTT container brought everything back.

**Reply by AlCalzone** [reply]
Maybe a stupid question, but did you try just feeding the `..._ota__TargetZwave__.exe` files directly into the z2m updater?
I haven't looked at them in detail but just by the directory contents, this looks a ton like Aeotec's updater tool, which we can extract the firmware updates from (usually).

**Reply by ChrisLizon** [reply]
Actually I did, I forgot to paste that. I just did it again to get the error message. 

```
2022-01-17 15:17:39.332 INFO ZWAVE: Calling api beginFirmwareUpdate with args: [
6,
'DZ1KD_01_20_ota__TargetZwave__.exe',
<Buffer 4d 5a 90 00 03 00 00 00 04 00 00 00 ff ff 00 00 b8 00 00 00 00 00 00 00 40 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ... 719584 more bytes>,
0,
[length]: 4
]
2022-01-17 15:17:39.351 INFO ZWAVE: Unable to extract firmware from file: This does not appear to be a valid Aeotec updater (invalid firmware name)! (ZW0322) beginFirmwareUpdate undefined
```

**Reply by AlCalzone** [reply]
Ok, looks like I need to check them out.

**Reply by KevinCathcart** [reply]
> Ok, looks like I need to check them out.

I'm not sure if the older exe's were like this, but the code for Aeotec updates looks *almost* correct for the current 1.22/1.23 Leviton exe updaters (from https://hubconnecteddecorasmart.zendesk.com/hc/en-us/articles/4407521680155), but those have an added wrinkle of the firmware being encypted with DES, and slightly different usage of the 256 byte area after the firmware.

For decryption, the code used is:
```c#
flashData = DESEncrypt.Decrypt(array2, "!@#$%^&*");
```
This is classic DES (not 3DES), in CBC mode with PKCS7 padding.  The UTF8 (ASCII) representation of the string passed second is used as both the IV, and the key. 

As mentioned, the 256 byte buffer is a little different. Instead of a filename, it has this format:
> 16 byte manufactureId (big endian)
> 16 byte productType (big endian)
> 16 byte productId (big endian)
> 16 byte versionId (big endian)
> 1 byte targetIndex
> 1 byte string length
> N bytes UTF8 (in practice ascii) string (brand)
> padded with null bytes

The existing code expects a non-length prefixed printable ASCII string, possibly preceded by a single byte target index, but obviously that won't work withthe above format. Which is probably for the best, since the existing code also does not know to decrypt the bytes, either.

Hope this helps.

**Comment by kaijk** [comment]
Leviton will provide links to the OTA FW files themselves. Contact support@decorasmartsupport.zendesk.com

Here are the links they sent me in late November for the then current DZ15S (v1.23) and DZ6HD (v1.22) FW files.  Note, these were more recent versions than those provided in the then current Aeotec executable...  I have installed them both with no issues using zwjs2mqtt.

DZ6HD: https://decorasmartsupport.zendesk.com/attachments/token/MFT4C61yHdtG43RDv3QkVHANg/?name=DZ6HD_v1_22.ota
DZ15S: https://decorasmartsupport.zendesk.com/attachments/token/laPYbWOXZISNjsA7Vw8sEESCw/?name=DZ15S_v1_23.ota

**Reply by ChrisLizon** [reply]
Thanks for pointing that out. While that's good to know, talk about terrible post-sales support. I'm going to use that link to update my DZ15S.  I may see what the latest available firmware is for the DZ1KD and DZPD3 while I'm at it.

**Reply by ChrisLizon** [reply]
Looks like even Leviton can't keep their stuff straight.

> Hello,
> 
> I was looking to update my Decora Z-wave devices to try to fix some stability issues I was having with my Z-wave devices. I had no luck using the download available and my Aeotec Z-stick gen 5. I purchased the Homeseer Z-Flash package which has update version 1.20, but I was made aware that there may be some newer firmware files available.
> 
> Can you please confirm the latest available firmware for the following devices, and send me the files?
> 
> DZPA1
> DZ1KD
> DZ15S
> DZPD3
> 
> Thank you very much.
> 
> Chris Lizon

And the response:

> 
> Good Morning,
> 1.20 is the latest firmware version for all of those devices.
> 
> Best regards,
> Andrew

@kaijk  not sure how you would feel about me replying back with those links...

**Comment by kaijk** [comment]
I don't have a problem with your referencing the links to the ota files in
a reply to their tech support.  I believe 1.20 is the latest FW  that's
included in the FW update executable, but there are demonstrably more
recent FW files available (at least for the DZ15S and DZ6HD).

**Comment by danieltwagner** [comment]
I've just received the following v1.22 OTA files from Leviton support for
- DZ6HD [DZ6HD_v1_22.ota.zip](https://github.com/zwave-js/node-zwave-js/files/8854544/DZ6HD_v1_22.ota.zip)
- DZ15S [DZ15S_v1_22.ota.zip](https://github.com/zwave-js/node-zwave-js/files/8854545/DZ15S_v1_22.ota.zip)
- DZPA1 [DZPA1_v1_22.ota.zip](https://github.com/zwave-js/node-zwave-js/files/8854546/DZPA1_v1_22.ota.zip)
- DZPD3 [DZPD3_v1_22.ota.zip](https://github.com/zwave-js/node-zwave-js/files/8854547/DZPD3_v1_22.ota.zip)

I came across this issue when looking for the files online so thought I'd post them here for others to find as well.

**Reply by ChrisLizon** [reply]
Thanks for sharing! My devices have generally been in the state of "If it ain't broke don't fix it" But these may be helpful if anyone else is looking for them.

**Comment by bpottle1** [comment]
Does anyone know what the changes are in these newest firmware? I have 1.20 on my switches and dimmers right now but I've been trying to see if there are any updates. Specifically I'm hoping they add multi-tap/central scene control and also fix the occasional weird connection issue I see where the switch stops updating state.

**Comment by danieltwagner** [comment]
The reason I updated was because I had connectivity issues (lack of updates, maybe lost messages) which improved, and I also noticed the state reporting back to the hub got a lot faster. I turn on another light when a DZ6HD changes state and that used to take around a second or so and is now basically instant.

**Reply by bpottle1** [reply]
Good to know. I've been having a couple issues like that too so hopefully this at least helps those. Thanks!

**Comment by escline** [comment]
Anybody have the updated file for the DZ1KD? I'm running v1.13 and have an issue where using a DD00R-DL 3way switch connected to the DZ1KD does not update Zwave, even though the lights update. So wondering if a firmware update would fix. I emailed them, but thought I'd also ask if anyone received the file already.

**Reply by escline** [reply]
I went ahead and purchased the HomeSeer Z-Wave update software. It does include both the 1.20 and 1.22 firmware for the DZ1KD. If Leviton responds to my query for the firmware with a link, I'll post it here.

And v1.22 did fix my issue with the 3 way switch not triggering event updates over Z-Wave.

**Comment by escline** [comment]
I did receive the ota file, in case anyone needs it:
[DZ1KD_v1_22.ota.zip](https://github.com/zwave-js/node-zwave-js/files/9074430/DZ1KD_v1_22.ota.zip)

**Comment by krehman** [comment]
Hi @ChrisLizon, just came across this post about updating the firmware. 

I am looking to update my DZ6HD dimmers and ordered the Zooz ZST10 700 stick. I just realized from your post that this is not supported by the Leviton updater tool provided in [Can Leviton Z-Wave Products Firmware Be Updated?](https://hubconnecteddecorasmart.zendesk.com/hc/en-us/articles/4407521680155-Can-Leviton-Z-Wave-Products-Firmware-Be-Updated-). Does the Leviton software (Leviton Smart Z-Wave OTA Tools) not pick-up the stick or there are other issues you faced? I have a Vera Edge hub in which I plan to add this ZST10 stick as a secondary controller as indicated in their manual, [Decora Smart Z-Wave OTA Firmware Updater.pdf](https://hubconnecteddecorasmart.zendesk.com/hc/en-us/article_attachments/7817982826523/Decora_Smart_Z-Wave_OTA_Firmware_Updater.pdf).

Any help would be appreciated or probably just stick out for the HomeSeer stick.

**Reply by kaijk** [reply]
Presumably you are running zwave-js under Homeassistant.  Grab the firmware file linked above on June 7 and use HA or zwjs2mqtt to apply it to your Dimmer.  Easy.

**Reply by krehman** [reply]
Thanks, no I am not using HA at all. Using homebridge with vera plugin.

**Reply by krehman** [reply]
An update for anyone who may stumble across this.

I was able to successfully update the Leviton switches using the Leviton updater tool provided in [Can Leviton Z-Wave Products Firmware Be Updated?](https://hubconnecteddecorasmart.zendesk.com/hc/en-us/articles/4407521680155-Can-Leviton-Z-Wave-Products-Firmware-Be-Updated-). I used a Zooz ZST10 700 stick.

I had to install [Simplicity Studio](https://www.silabs.com/developers/z-wave-800-series) which provided the drivers for the ZST10 from there it was just following the [Decora Smart Z-Wave OTA Firmware Updater.pdf](https://hubconnecteddecorasmart.zendesk.com/hc/en-us/article_attachments/7817982826523/Decora_Smart_Z-Wave_OTA_Firmware_Updater.pdf).

Note - I have a Vera Edge hub, and as instructed, I joined the network and updated the nodes. Overall, surprisingly a pretty painless process and quite a few switches updated quickly (~5 mins) whereas some took (~45mins). I am guessing due to the distance.

Also, for the latest firmware to show up on the Vera Edge, I had to update the nodes through **Devices** > **Advanced** > **Commands** > **Configure node right now**

![image](https://user-images.githubusercontent.com/6234445/198093192-c5ba422f-c545-4f0e-b6b2-a73adb292c24.png)

**Screenshots**
![IMG_6211](https://user-images.githubusercontent.com/6234445/198093334-9f0fc78e-10ae-47c1-a154-66f5567b1c79.jpg)
![IMG_6216](https://user-images.githubusercontent.com/6234445/198093337-6d1aaf60-a9cd-4225-a524-8b0fc1ba7105.jpg)
![IMG_6217](https://user-images.githubusercontent.com/6234445/198093340-dc17933b-9415-4a72-b1f7-bc55502c90f7.jpg)

EDIT: added screenshots

**Comment by nbrewk** [comment]
This is a summary post for anyone coming across this discussion on Google, as this page was coming up as the first result for Leviton Z-wave Device updates. I was able to complete the updates using Hubitat using the linked files in this thread.

After a little more research, I found this Leviton support site links all the proper current files (at the bottom of the page) without the Z-wave update wrapper:
https://hubconnecteddecorasmart.leviton.com/hc/en-us/articles/7151993826075-Updating-Decora-Smart-Z-Wave-Using-Home-Assistant

**Reply by ChrisLizon** [reply]
That's awesome that Leviton has updated their documentation to include Home Assistant!

**Comment by kaijk** [comment]
They're a (the first) "Works with Home Assistant" partner
